### PR TITLE
Speed up import of pyomo.environ

### LINKED ISSUE #1152: Performance suggestion: Pyomo import performance
Importing pyomo can take a long time depending on what packages are installed.

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
The time it takes to `import pyomo.environ` has been steadily increasing.  After a bit of profiling (see the exceptionally useful visualization package [tuna](https://pypi.org/project/tuna/)), we tracked it down to a number of packages that are slow to import; namely:
- sympy (0.68s)
- pandas (0.52s)
- IPython (0.32s)
- pkg_resources (0.31s)
- openpyxl (0.28s)
- networkx (0.18s)
- pint (0.07s)
- yaml (0.04s)

This PR leverages the `attempt_import()` logic to perform "deferred" imports of key dependencies.  In this way, the dependencies will only be imported the first time that code attempts to use the module (or checks the status of the "available" flag).

This PR reduces `import pyomo.environ` from ~4s to roughly 0.5s.  Of the remaining import time:
- 25% is importing numpy: we unconditionally check if numpy is available and register the numpy types with the expression system.  A subsequent PR will look at the impact of deferring numpy registration until the first time one of its types appears in an expression
- 8% is `pyutilib.misc`, the bulk of that is `tempfile`.  Currently, most workflows involve creating at least one temp file, so avoiding this import here isn't a huge win.  However, as we move away from file-based solver interfaces, we should look at avoiding these imports.
- 4% is importing `pyomo.pysp.ph` (I believe simply because it is a 5k line file)
- 4% is importing Pyro4 (the import is hard-coded into `pyutilib.pyro.util`, and deferring that import will be more challenging.
- 2% is `pyomo.scripting.util`
- 1.7% is `pyomo.pysp.util.config`

For concrete impact, running `test.pyomo pyomo.core` reduces from over 60 seconds to 25 seconds.

## Changes proposed in this PR:
- overhaul of `attempt_import` to support deferred imports
- deferred import of key packages: `yaml`, `pympler`, `numpy`, `scipy`, `networkx`, `pandas`, `dill`,
- modify how we interact with the `pyutilib.excel` spradsheet interface

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by blnicho** [comment]
> This PR reduces import pyomo.environ from ~4s to roughly 0.5s.

Do these times factor in the changes you've also made to PyUtilib to defer imports?

**Comment by jsiirola** [comment]
Yes: the net import time reduction includes this PR plus a bit of work in PyUtilib.  The changes to pyutilib were mostly to avoid importing things like `openpyxl`, and were necessary to avoid importing `yaml` (which was imported in about a dozen places spread across both projects).

**Comment by jsiirola** [comment]
I believe this is ready to review.  The only "failing" check is coverage, and it's hard to tell from the codecov report where the coverage really decreased.

File: pyomo/dataportal/plugins/xml_table.py [hunk 1]
```diff
@@ -9,14 +9,29 @@
...
 
+def _xml_importer():
+    try:
+        from lxml import etree
+        return etree
+    except ImportError:
+        pass
+
+    try:
+        # Python 2.5+
```
Comment (blnicho) [inline pyomo/dataportal/plugins/xml_table.py]: Did you mean to type Python 2.5 here or is this a typo?

Comment (jsiirola) [reply pyomo/dataportal/plugins/xml_table.py]: Yes.  I meant Pyomo 2.5 (that syntax works as far back as python 2.5, but no farther).  There is reference documentation that also includes imports for 2.4 and earlier, but I intentionally did not include those imports here.

File: pyomo/common/dependencies.py [hunk 2]
```diff
@@ -37,6 +38,130 @@ def generate_import_warning(self, logger='pyomo.common'):
         logging.getLogger(logger).warning(
             self._error_message_)
 
+class DeferredImportModule(object):
+    """Dummy object that serves as a module placeholder until the first time
+    getattr is called, at which point it imports the module and returns
+    the module attribute.
+
+    This object is returned by attempt_import() in liu of the module in
```
[the remark this answers is not in the document]
Comment (jsiirola) [reply pyomo/common/dependencies.py]: fixed.

**Comment by blnicho** [comment]
@jsiirola I merged the ParmEst PR before I realized there were conflicts with this one. Once we resolve the conflicts I think this is ready to merge.

**Pull request merged**